packages/eas-cli/src/credentials/ios/actions/SetupBuildCredentials.ts

```diff
@@ -31,6 +35,31 @@ interface IosAppBuildCredentials {
   };
 }
 
+function isNativeAssociatedDomainsDefined(projectDir: string): boolean | null {
+  try {
+    const entitlementsPath = IOSConfig.Paths.getEntitlementsPath(projectDir);
+    if (entitlementsPath) {
+      const entitlements = plist.parse(fs.readFileSync(entitlementsPath, 'utf8'));
+      return !!entitlements['com.apple.developer.associated-domains'];
+    }
+  } catch {}
+  return null;
+}
+
+function isAssociatedDomainsDefined(projectDir: string, exp: Pick<ExpoConfig, 'ios'>) {
+  // Get the native entitlements file and determine if associated domains is defined in it.
+  // If we don't enable this capability xcode will fail to build:
+  // error: Provisioning profile "*[expo] com.bacon.myapp AppStore 2021-05-04T22:29:48.816Z" doesn't support the Associated Domains capability.
+  let isAssociatedDomainsDefined = isNativeAssociatedDomainsDefined(projectDir);
+  if (isAssociatedDomainsDefined == null) {
+    // Determine if the app.json has `ios.associatedDomains` or equivalent entitlements field defined.
+    isAssociatedDomainsDefined =
+      !!exp.ios?.associatedDomains ||
+      !!exp.ios?.entitlements?.['com.apple.developer.associated-domains'];
+  }
+  return isAssociatedDomainsDefined;
+}
```

Comment:
Maybe move this logic outside of the file, it's at a different level of abstraction than other functions in this file. 
I'd create `CapabilityUtils.ts` or similar in the same directory.